modules/ControlGUI/src/uimoosinterface.cpp

```diff
@@ -0,0 +1,387 @@
+        //OK we are going to wait for a conenction to be established
+        // this is a little harsh but it saves derived classes having to
+        // hold off connecting to the server until ready
+        // but we will only hang around for 1 second...
+        // so it is possible that notifies will fail...but very unlikely
+        // note this is not a hack! just being helpful. Ths success of an
+        // application is NOT dependent on this
+        int t = 0;
+        int dT = 50;
+        while(!m_Comms.IsConnected())
+        {
+            MOOSPause(dT);
+            t+=dT;
+            if(t>5000)
+                break;
+        }
+        //give iostream time to write comms start details up to screen..this is not really necessary
+        //as the code is thread safe...it is aesthetic only
+        MOOSPause(500);
+    }
+
+
+    /** let derivatives do stuff before execution*/
+    if(!OnStartUpPrepare())
+    {
+        MOOSTrace("Derived OnStartUpPrepare() returned false... Quitting\n");
+        return false;
+    }
+
+    if(!OnStartUp())
+    {
+        MOOSTrace("Derived OnStartUp() returned false... Quitting\n");
+        return false;
+    }
+
+
+    if(!OnStartUpComplete())
+    {
+        MOOSTrace("Derived OnStartUpComplete() returned false... Quitting\n");
+        return false;
+    }
+
+    if(m_CommandLineParser.GetFlag("--moos_configuration_audit"))
+    {
+        PrintSearchedConfigurationFileParameters();
+        return true;
+    }
+
+
+    DoBanner();
+    connect(&iterateTimer, &QTimer::timeout, this, &UIMoosInterface::doMOOSWork);
+    iterateTimer.start(1000.0/m_dfFreq);
+    currentFrequency = m_dfFreq;
+    return true;
+}
+
+bool UIMoosInterface::doMOOSWork()
+{    /****************************  THE MAIN MOOS APP LOOP **********************************/
+    m_dfFreq = -1; //Set this to -1 in order to override the MOOS Sleep function
+    bool bOK = DoRunWork();
+    if(m_bQuitOnIterateFail && !bOK){
+        m_Comms.Close();
+        emit workFinished();
+        return MOOSFail("MOOSApp Exiting as requested");
+    }
+    /***************************   END OF MOOS APP LOOP ***************************************/
+    return true;
+}
+
+void UIMoosInterface::receiveZeta(QString zeta)
+{
+    Notify("Zeta_Cmd", zeta.toStdString());
+}
+
+void UIMoosInterface::receiveStateCMD(EnumDefs::VehicleStates state, QString id, int maxSpeed)
+{
+    if(state != EnumDefs::VehicleStates::SWARMMODE && id != QString("All"))
+        Notify(id.toStdString()+"_Change_State", "State="+QString::number(int(state)).toStdString()+", id="+id.toStdString());
+    else
+        Notify("Change_State", "State="+QString::number(int(state)).toStdString());
+}
+
+void UIMoosInterface::receiveSpeed(QString id, bool forward, bool reverse, bool left, bool right, int speed)
+{
+    int speedSign = 1;
+    int curv = 0;
+    //Forward
+    if(forward and ((not left and not right) or (left and right))){
```

Comment:
I know the words and work, but I hate them. No need to change them but I hate them.